packages/vite/src/node/ssr/ssrModuleLoader.ts

```diff
@@ -324,3 +350,58 @@ function proxyESM(mod: any) {
 function isPrimitive(value: any) {
   return !value || (typeof value !== 'object' && typeof value !== 'function')
 }
+
+/**
+ * Vite converts `import { } from 'foo'` to `const _ = __vite_ssr_import__('foo')`.
+ * Top-level imports and dynamic imports work slightly differently in Node.js.
+ * This function normalizes the differences so it matches prod behaviour.
+ */
+function analyzeImportedModDifference(
+  mod: any,
+  filePath: string,
+  rawId: string,
+  metadata?: SSRImportMetadata,
+) {
+  // No normalization needed if the user already dynamic imports this module
+  if (metadata?.isDynamicImport) return
+  // If file path is ESM, everything should be fine
+  if (isFilePathESM(filePath)) return
```

Comment:
Given that `isFilePathESM` calls `lookupFile` and invokes multiple fs calls, maybe we should add a cache to avoid hurting performance.

Reply:
Good catch! A lot of this was actually written as a worst-case before, but seems like it's hotter now. I'll try to optimize this part.